flutter/lib/src/span_frame_metrics_collector.dart

```diff
@@ -0,0 +1,212 @@
+
+import 'frame_callback_handler.dart';
+import 'native/sentry_native.dart';
+
+@internal
+class SpanFrameMetricsCollector implements PerformanceContinuousCollector {
+  final SentryFlutterOptions options;
+  final FrameCallbackHandler? _frameCallbackHandler;
+  final SentryNative? _native;
+
+  /// Stores frame timestamps and their durations in milliseconds.
+  /// Keys are frame timestamps, values are frame durations.
+  final frameDurations = SplayTreeMap<DateTime, int>();
+
+  /// Stores the spans that are actively being tracked.
+  /// After the frames are calculated and stored in the span the span is removed from this list.
+  final activeSpans = <ISentrySpan>[];
+
+  bool get isTrackingPaused => _isTrackingPaused;
+  bool _isTrackingPaused = true;
+
+  bool get isTrackingRegistered => _isTrackingRegistered;
+  bool _isTrackingRegistered = false;
+
+  final _stopwatch = Stopwatch();
+
+  SpanFrameMetricsCollector(this.options,
+      {FrameCallbackHandler? frameCallbackHandler, SentryNative? native})
+      : _frameCallbackHandler =
+            frameCallbackHandler ?? DefaultFrameCallbackHandler(),
+        _native = native ?? SentryFlutter.native;
+
+  @override
+  void onSpanStarted(ISentrySpan span) {
+    if (span is NoOpSentrySpan || !options.enableFramesTracking) {
+      return;
+    }
+
+    activeSpans.add(span);
+    startFrameTracking();
+  }
+
+  @override
+  Future<void> onSpanFinished(ISentrySpan span, DateTime endTimestamp) async {
+    if (span is NoOpSentrySpan || !options.enableFramesTracking) {
+      return Future.value();
+    }
+
+    await recordSpanFrameMetrics(span, endTimestamp);
+
+    if (activeSpans.isEmpty) {
+      clear();
+    } else {
+      final oldestSpan = activeSpans.first;
+      frameDurations.removeWhere((key, value) {
+        return key.isBefore(oldestSpan.startTimestamp);
+      });
+    }
+  }
+
+  /// Calls [WidgetsBinding.instance.addPersistentFrameCallback] which cannot be unregistered
+  /// and exists for the duration of the application's lifetime.
+  ///
+  /// Stopping the frame tracking means setting [isTrackingPaused] is `true`
+  /// to prevent actions being done when the frame callback is triggered.
+  void startFrameTracking() {
+    _isTrackingPaused = false;
+
+    if (!_isTrackingRegistered) {
+      _frameCallbackHandler?.addPersistentFrameCallback(measureFrameDuration);
+      _isTrackingRegistered = true;
+    }
+  }
+
+  /// Records the duration of a single frame and stores it in [frameDurations].
+  ///
+  /// This method is called for each frame when frame tracking is active.
+  void measureFrameDuration(Duration duration) async {
+    if (_isTrackingPaused) return;
+
+    if (!_stopwatch.isRunning) {
+      _stopwatch.start();
+    }
+
+    await _frameCallbackHandler?.endOfFrame;
+
+    final frameDuration = _stopwatch.elapsedMilliseconds;
+    // ignore: invalid_use_of_internal_member
+    frameDurations[getUtcDateTime()] = frameDuration;
```

Comment:
just for consistency
```suggestion
    frameDurations[options.clock()] = frameDuration;
```